src/Measurement_Observation/haskell_model.rst

```diff
@@ -0,0 +1,129 @@
+For example:
+
+1. we don't have clear procedure boundaries with call/return instruction pairs
+   (we use tail calls, i.e. jump instructions)
+
+2. we don't use the so-called C stack in the usual way. I.e. we don't use usual
+   stack registers (e.g. rsp/rbp on x86-64) as stack registers.
+
+3. GHC uses its own calling convention
+
+In addition, lazy evaluation makes control-flow and memory usage difficult to
+understand.
+
+In this chapter we describe the compilation pipeline and the execution model.
+For each stage of the pipeline we list what can be measured at runtime.
+
+
+Compilation pipeline overview
+-----------------------------
+
+GHC compiles Haskell codes in several stages, each with their own intermediate
+representation:
+
+Haskell: the functional language we love with its bazillion of extensions.
+
+Core: a simple and explicitly typed functional language. Types are basically
+Haskell types. Type-class dictionaries are passed explicitly as records,
+type-applications are used everywhere needed, coercions (proofs that we can
+convert a type into another) are first class values, etc.
+
+STG: a functional language closer to the execution model. It's still typed but
+with primitive types. E.g. it tracks if a value is a heap object or an
+unboxed word (unboxed = doesn't have its own heap object) but not the Haskell
+type of the heap object. Complex forms are lowered to simpler ones (e.g. unboxed
+sums are lowered into unboxed tuples); this is called unarisation.
+
+Additionally every heap allocation is made explicit with a let-binding. For
+example instead of ``foo (MkBar x y)`` you have ``let b = MkBar x y in foo b``.
+This is called `A-normal form<https://en.wikipedia.org/wiki/A-normal_form>`_.
+During code generation from STG we know that all functions are applied to
+"simple" arguments only (variables, constants, etc.).
+
+Following the approach pioneered with "super-combinators", every top-level STG
+binding is then compiled into imperative code. The idea is that executing this
+imperative code has the same result as interpreting the functional code. For
+example, ``let b = MkBar x y in foo b`` is basically compiled to:
+
+.. code-block:: C
+
+  b <- allocate heap object for `MkBar x y`
+  evaluate foo
+  apply `foo` to `b`
+
+Different GHC backends use different imperative representations. The interpreter
+uses ByteCode, the JavaScript backend uses some JavaScript-like form, all other
+backends use Cmm.
+
+Cmm: an imperative language that looks like LLVM IR. It supports expressions,
+statements, and it abstracts over machine primops, machine registers, stack
+usage, and calling conventions. A pass performs register assignment and stack
+layout for the target architecture. Then assembly code (textual) is generated
+for the target architecture and an external assembler program generates machine
+code (binary) for it. Finally an external linker is used to transform the
+resulting code objects into a single executable or library.
```

Comment:
We're missing a paragraph that states:

- what information can I get by reading Cmm that I cannot get by reading STG and Core?
- and then a link: 
```
`:ref:`Reading Cmm <Reading Cmm>` 
```
to the reading cmm chapter.